smithy-model/src/main/java/software/amazon/smithy/model/knowledge/PropertyBindingIndex.java

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.model.knowledge;
+
+import java.lang.ref.WeakReference;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.shapes.MemberShape;
+import software.amazon.smithy.model.shapes.OperationShape;
+import software.amazon.smithy.model.shapes.ResourceShape;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.shapes.ShapeId;
+import software.amazon.smithy.model.traits.NestedPropertiesTrait;
+import software.amazon.smithy.model.traits.NotPropertyTrait;
+import software.amazon.smithy.model.traits.PropertyTrait;
+
+/**
+ * Index of member shape to associated resource property information.
+ */
+public final class PropertyBindingIndex implements KnowledgeIndex {
+    private final WeakReference<Model> model;
+    private final OperationIndex operationIndex;
+    private final Set<ShapeId> notPropertyMetaTraitSet;
+    private final Map<ShapeId, Boolean> memberShapeDoesNotRequireProperty = new HashMap<>();
```

Comment:
Maps of things to booleans can be replaced with sets, where the presence in the set indicates a `true` state.

Reply:
Need this due to a difference between a ShapeId having a mapping to a boolean true or false value, and ShapeId not being a key in the map at all.